--- a/google-modules/soc/gs/drivers/soc/google/pkvm-s2mpu/common/hyp/s2mpu.c
+++ b/google-modules/soc/gs/drivers/soc/google/pkvm-s2mpu/common/hyp/s2mpu.c
@@ -18,6 +18,7 @@
 #include <nvhe/trap_handler.h>
 #include "io-mpt-s2mpu.h"
 
+#ifdef CONFIG_MODULES
 #ifdef memset
 #undef memset
 #endif
@@ -26,6 +27,7 @@
 #ifdef memcpy
 #undef memcpy
 #endif
+#endif
 
 #define PA_MAX				((phys_addr_t)SZ_1G * NR_GIGABYTES)
 
@@ -58,6 +60,7 @@ struct s2mpu_drv_data {
 };
 
 static const struct s2mpu_mpt_ops *mpt_ops;
+#ifdef CONFIG_MODULES
 static const struct pkvm_module_ops *mod_ops;
 
 /* Abstraction for function used from module ops. */
@@ -85,6 +88,9 @@
 {
 	return CALL_FROM_OPS(memcpy, dst, src, count);
 }
+#else
+#define mod_ops NULL
+#endif
 
 static struct mpt host_mpt;
 
@@ -657,6 +663,7 @@
 	.ops = &pkvm_sysmmu_sync_ops,
 };
 
+#ifdef CONFIG_MODULES
 /* Validate that module ops used within this driver are not null. */
 static bool validate_mod_ops(const struct pkvm_module_ops *ops)
 {
@@ -682,4 +689,5 @@
 
 	return 0;
 }
+#endif
 
