--- a/google-modules/soc/gs/drivers/soc/google/pkvm-s2mpu/common/hyp/io-mpt-s2mpu.c
+++ b/google-modules/soc/gs/drivers/soc/google/pkvm-s2mpu/common/hyp/io-mpt-s2mpu.c
@@ -8,6 +8,7 @@
 #define SMPT_NUM_TO_BYTE(x)		((x) / SMPT_GRAN / SMPT_ELEMS_PER_BYTE(MPT_PROT_BITS))
 #define BYTE_TO_SMPT_INDEX(x)		((x) / SMPT_WORD_BYTE_RANGE(MPT_PROT_BITS))
 
+#ifdef CONFIG_MODULES
 #ifdef memset
 #undef memset
 #endif
@@ -29,6 +30,7 @@ static const struct pkvm_module_ops *mod_ops;
 #undef kvm_flush_dcache_to_poc
 #endif
 #define kvm_flush_dcache_to_poc(in, sz)	CALL_FROM_OPS(flush_dcache_to_poc, in, sz)
+#endif
 
 
 static inline int pte_from_addr_smpt(u32 *smpt, u64 addr)
@@ -286,6 +288,8 @@ static const struct s2mpu_mpt_ops this_ops = {
 
 const struct s2mpu_mpt_ops *s2mpu_get_mpt_ops(const struct pkvm_module_ops *in_mod_ops)
 {
+#ifdef CONFIG_MODULES
 	mod_ops = in_mod_ops;
+#endif
 	return &this_ops;
 }
--- a/google-modules/soc/gs/drivers/soc/google/pkvm-s2mpu/common/hyp/s2mpu.c
+++ b/google-modules/soc/gs/drivers/soc/google/pkvm-s2mpu/common/hyp/s2mpu.c
@@ -20,6 +20,7 @@
 #include <nvhe/trap_handler.h>
 #include "io-mpt-s2mpu.h"
 
+#ifdef CONFIG_MODULES
 #ifdef memset
 #undef memset
 #endif
@@ -27,6 +28,7 @@
 #ifdef memcpy
 #undef memcpy
 #endif
+#endif
 
 #define PA_MAX				((phys_addr_t)SZ_1G * NR_GIGABYTES)
 
@@ -58,6 +60,7 @@ struct s2mpu_drv_data {
 };
 
 static const struct s2mpu_mpt_ops *mpt_ops;
+#ifdef CONFIG_MODULES
 static const struct pkvm_module_ops *mod_ops;
 
 /* Abstraction for function used from module ops. */
@@ -85,6 +88,9 @@ void *memcpy(void *dst, const void *src, size_t count)
 {
 	return CALL_FROM_OPS(memcpy, dst, src, count);
 }
+#else
+#define mod_ops NULL
+#endif
 
 static struct mpt host_mpt;
 
@@ -657,6 +663,7 @@ struct pkvm_iommu_driver pkvm_sysmmu_sync_driver = (struct pkvm_iommu_driver){
 	.ops = &pkvm_sysmmu_sync_ops,
 };
 
+#ifdef CONFIG_MODULES
 /* Validate that module ops used within this driver are not null. */
 static bool validate_mod_ops(const struct pkvm_module_ops *ops)
 {
@@ -682,4 +689,5 @@ int s2mpu_hyp_init(const struct pkvm_module_ops *ops)
 
 	return 0;
 }
+#endif
 
--- a/google-modules/soc/gs/drivers/soc/google/pkvm-s2mpu/common/pkvm_s2mpu.c
+++ b/google-modules/soc/gs/drivers/soc/google/pkvm-s2mpu/common/pkvm_s2mpu.c
@@ -427,11 +427,13 @@ static int s2mpu_driver_register(struct platform_driver *driver)
 	}
 
         if (is_protected_kvm_enabled()) {
+#ifdef CONFIG_MODULES
 		ret = pkvm_load_el2_module(__kvm_nvhe_s2mpu_hyp_init, &token);
 		if (ret) {
 			pr_err("Failed to load s2mpu el2 module: %d\n", ret);
 			return ret;
 		}
+#endif
 
 		ret = pkvm_iommu_s2mpu_init(token);
 		if (ret) {
--- a/google-modules/soc/gs/drivers/soc/google/pkvm-s2mpu/common/s2mpu.c
+++ b/google-modules/soc/gs/drivers/soc/google/pkvm-s2mpu/common/s2mpu.c
@@ -11,9 +11,14 @@
 
 /* Token of S2MPU driver, token is the load address of the module. */
 static u64 token;
+
+#ifdef CONFIG_MODULES
 /* For an nvhe symbol loaded as a module, get the hyp address of it. */
 #define ksym_ref_addr_nvhe(x)	((u64)(&kvm_nvhe_sym(x)) - \
 				 (u64)THIS_MODULE->arch.hyp.text.start + token)
+#else
+#define ksym_ref_addr_nvhe(x)	kvm_nvhe_ref_addr(&kvm_nvhe_sym(x))
+#endif
 
 
 extern struct pkvm_iommu_driver kvm_nvhe_sym(pkvm_sysmmu_sync_driver);
